### PR TITLE
doc: add disk benchmarking and cache tuning recommendations

### PR DESCRIPTION
**Pull request by dvanders**
Fixes: https://tracker.ceph.com/issues/53161
Signed-off-by: Dan van der Ster <daniel.vanderster@cern.ch>

### CODE REVIEW

**Comment by dvanders** [comment]
@vitalif @k0ste does it look okay?
(The earlier HDD and SSD text also needs to be brought up to date, but that's for another PR).

File: doc/start/hardware-recommendations.rst [hunk 1]
```diff
@@ -236,6 +236,48 @@ costs.  Some RAID HBAs can be configured with an IT-mode "personality".
...
+these types of devices can be toggled between two caching modes -- a
+volatile cache flushed safely with fsync, or a non-volatile cache written
+synchronously. Counterintuitively, these two modes are selected by either
+"enabling" or "disabling" the device cache.
+
+The cache configuration can be queried with ``hdparm -W`` or by reading the
+values in ``/sys/devices/scsi_disk/*/cache_type``. The default configuration
+-- normally cache enabled / "write back" -- is often not optimal. The
+alternative -- cache disabled / "write through" -- is often observed to
+dramatically increase random write performance on both modern SSDs and HDDs.
```
Comment (markhpc) [inline doc/start/hardware-recommendations.rst:264]: nit: We may want to consider switching the wording from "is often" to "may" unless we can provide some concrete examples here to go along with it.

File: doc/start/hardware-recommendations.rst [hunk 2]
```diff
@@ -236,6 +236,48 @@ costs.  Some RAID HBAs can be configured with an IT-mode "personality".
...
+
+Users are therefore encouraged to benchmark new devices with ``fio`` as
+described earlier and persist the optimal cache configuration for their
+devices.
+
+.. tip:: Modifying cache_type sets the device cache status; on some distros
```
Comment (markhpc) [inline doc/start/hardware-recommendations.rst:270]: Which distros?  When should a user do this vs the udev rule below?  This should be fleshed out a little more.

File: doc/start/hardware-recommendations.rst [hunk 3]
```diff
@@ -236,6 +236,48 @@ costs.  Some RAID HBAs can be configured with an IT-mode "personality".
...
+-----------
+
+Most enterprise-grade SSDs and HDDs include power loss protection features
+which employ multi-level caches to speed up direct random writes. Many of
+these types of devices can be toggled between two caching modes -- a
+volatile cache flushed safely with fsync, or a non-volatile cache written
+synchronously. Counterintuitively, these two modes are selected by either
+"enabling" or "disabling" the device cache.
+
+The cache configuration can be queried with ``hdparm -W`` or by reading the
```
[the remark this answers is not in the document]
Comment (dvanders) [reply doc/start/hardware-recommendations.rst]: Good idea, I've added examples to getting and setting.

File: doc/start/hardware-recommendations.rst (hunk 1 above)
Comment (dvanders) [reply doc/start/hardware-recommendations.rst]: done

File: doc/start/hardware-recommendations.rst (hunk 2 above)
Comment (dvanders) [reply doc/start/hardware-recommendations.rst]: I've fleshed it out, better now?

**Comment by dvanders** [comment]
@k0ste @markhpc better now?

**Comment by k0ste** [comment]
Some tests under backfill load (TOSHIBA MG08ACA16TE)
Drive utilization is decreased
Screen Shot 2021-11-15 at 12 18 26
OSD commit latency is decreased
Screen Shot 2021-11-15 at 12 18 01
Drive IOPS is increasing
Screen Shot 2021-11-15 at 12 18 17

**Comment by dvanders** [comment]
Commit Latency on Toshiba MG07ACA14TE
(RBD cluster, normal user workload, no backfilling ongoing)

image

File: doc/start/hardware-recommendations.rst [hunk 4]
```diff
@@ -236,6 +236,98 @@ costs.  Some RAID HBAs can be configured with an IT-mode "personality".
...
+Write Caches
+------------
+
+Enterprise SSDs and HDDs normally include power loss protection features which
+use multi-level caches to speed up direct or synchronous writes.  These devices
+can be toggled between two caching modes -- a volatile cache flushed to
+persistent media with fsync, or a non-volatile cache written synchronously.
+
+These two modes are selected by either "enabling" or "disabling" the write
+cache.  When the volatile cache is enabled, Linux uses a device in "write back"
```
Comment (athanatos) [inline doc/start/hardware-recommendations.rst:260]: the write (volatile) cache

**Review by athanatos — APPROVED** [review]
One nit, looks good otherwise

**Comment by dvanders** [comment]
jenkins test needs-rebase

**Comment by dvanders** [comment]
jenkins retest this please

**Comment by plieven** [comment]
@dvanders it might be worth also mentioning the sdparm utility in the docs. It can handle both SATA and SAS devices at the same time. It can also handle multiple devices at once:

`sdparm --get WCE /dev/sd*`

`sdparm --clear WCE /dev/sd*`

**Comment by dvanders** [comment]
@plieven thanks I'll incorporate that.

For now I'm pausing this because we just found that the udev rule doesn't work on CentOS 7

**Comment by dvanders** [comment]
@zdover23 this is ready for merge

**Comment by zdover23** [comment]
jenkins retest this please

File: doc/start/hardware-recommendations.rst [hunk 5]
```diff
@@ -236,6 +236,119 @@ costs.  Some RAID HBAs can be configured with an IT-mode "personality".
...
+
+The default configuration (normally caching enabled) may not be optimal, and
+OSD performance may be dramatically increased in terms of increased IOPS and
+decreased commit_latency by disabling the write cache.
+
+Users are therefore encouraged to benchmark their devices with ``fio`` as
+described earlier and persist the optimal cache configuration for their
+devices.
+
+The cache configuration can be queried with ``hdparm``, ``smartctl`` or by
```
Comment (mykaul) [inline doc/start/hardware-recommendations.rst:271]:
```suggestion
The cache configuration can be queried with ``hdparm``, ``sdparm``, ``smartctl`` or by
```

**Comment by mykaul** [comment]
Just wondering, have you tried 'sdparm --clear=WCE=0' to turn off write caching?

**Comment by dvanders** [comment]
> Just wondering, have you tried 'sdparm --clear=WCE=0' to turn off write caching?

It's in the "tip" at the bottom of the commit.
We can document this way, but in my testing the kernel does not always change to write through mode after clearing WCE directly. So the echo "write through" approaches are more reliable, afaict.

**Comment by k0ste** [comment]
> Just wondering, have you tried 'sdparm --clear=WCE=0' to turn off write caching?
We use exactly this two rules:
```shell
ACTION=="add|change", SUBSYSTEM=="scsi_disk", ATTR{cache_type}="write through"
KERNEL=="sd*", ACTION=="add|change", SUBSYSTEM=="block", ATTRS{queue/rotational}=="1", ENV{DEVTYPE}=="disk", RUN+="/usr/bin/sdparm --clear=WCE /dev/%k"
```

**Pull request merged**